src/pileuputils.js

```diff
@@ -95,8 +86,95 @@ function getDifferingBasePairs(read: SamRead, reference: string): Array<BasePair
   return out;
 }
 
+type OpInfo = {
+  ops: Object[],
+  mismatches: BasePair[]
+}
+
+// Determine which alignment segment to render as an arrow.
+// This is either the first or last 'M' section, excluding soft clipping.
+function getArrowIndex(read: SamRead): number {
+  var i, op, ops = read.getCigarOps();
+  if (read.getStrand() == '-') {
+    for (i = 0; i < ops.length; i++) {
+      op = ops[i];
+      if (op.op == 'S') continue;
+      if (op.op == 'M') return i;
+      return -1;
+    }
+  } else {
+    for (i = ops.length - 1; i >= 0; i--) {
+      op = ops[i];
+      if (op.op == 'S') continue;
+      if (op.op == 'M') return i;
+      return -1;
+    }
+  }
+  return -1;
+}
+
+// Breaks the read down into Cigar Ops suitable for display
+function getOpInfo(read: SamRead, referenceSource: Object): OpInfo {
+  var ops = read.getCigarOps();
+
+  var range = read.getInterval(),
+      start = range.start(),
+      seq = read.getSequence(),
+      seqPos = 0,
+      refPos = start,
+      arrowIndex = getArrowIndex(read);
+
+  var result = [];
+  var mismatches = ([]: BasePair[]);
+  for (var i = 0; i < ops.length; i++) {
+    var op = ops[i];
+    if (op.op == 'M') {
+      var ref = referenceSource.getRangeAsString({
+        contig: range.contig,
+        start: refPos,
+        stop: refPos + op.length - 1
+      });
+      var mSeq = seq.slice(seqPos, seqPos + op.length);
+      mismatches = mismatches.concat(findMismatches(ref, mSeq, refPos));
+    }
+
+    result.push({
+      op: op.op,
+      length: op.length,
+      pos: refPos
+    });
+
+    switch (op.op) {
```

Comment:
(I realize it's CIGAR parsing, but it'd help to have it briefly documented here, anyway—your call though)